<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Motivating example for application of SNN to single cell cancer data</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single_cell_cancer</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/single_cell_cancer">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Motivating example for application of SNN to single cell cancer data</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2022-1-17</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-01-17
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>single_cell_cancer/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20211008code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20211008)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20211008code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20211008)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiusinglecellcancertreec5c86b1e3d085d748a7e4de5b0f531bc684d02d9targetblankc5c86b1a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/single_cell_cancer/tree/c5c86b1e3d085d748a7e4de5b0f531bc684d02d9" target="_blank">c5c86b1</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiusinglecellcancertreec5c86b1e3d085d748a7e4de5b0f531bc684d02d9targetblankc5c86b1a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/single_cell_cancer/tree/c5c86b1e3d085d748a7e4de5b0f531bc684d02d9" target="_blank">c5c86b1</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    analysis/.DS_Store
    Ignored:    code/.DS_Store

Untracked files:
    Untracked:  analysis/twogroup_binary_general.Rmd
    Untracked:  code/functions.R
    Untracked:  code/simulations.R
    Untracked:  code/simulations3_more.R
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep1.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep10.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep2.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep3.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep4.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep5.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep6.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep7.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep8.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_05_rep9.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep1.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep10.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep2.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep3.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep4.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep5.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep6.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep7.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep8.rds
    Untracked:  output/simulations3_more_XXt_fit1_sigma_1_rep9.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_hist.pdf
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep1.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep10.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep2.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep3.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep4.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep5.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep6.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep7.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep8.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_05_rep9.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_hist.pdf
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep1.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep10.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep2.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep3.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep4.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep5.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep6.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep7.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep8.rds
    Untracked:  output/simulations3_more_XXt_fit2_sigma_1_rep9.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep1.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep10.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep2.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep3.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep4.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep5.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep6.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep7.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep8.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_05_rep9.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep1.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep10.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep2.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep3.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep4.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep5.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep6.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep7.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep8.rds
    Untracked:  output/simulations3_more_XXt_fit3_sigma_1_rep9.rds
    Untracked:  output/simulations3_update_XXt_fit3_sigma_v1.Rds
    Untracked:  output/simulations3_update_XXt_fit3_sigma_v2.Rds

Unstaged changes:
    Modified:   code/simulations3_update.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/presentation.Rmd</code>) and HTML (<code>docs/presentation.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/single_cell_cancer/blob/c5c86b1e3d085d748a7e4de5b0f531bc684d02d9/analysis/presentation.Rmd" target="_blank">c5c86b1</a>
</td>
<td>
yushaliu
</td>
<td>
2022-01-17
</td>
<td>
Add a motivating example for illustration
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>We present a toy example which illustrates the characteristics observed in single cell cancer data, which demonstrate strong inter-patient heterogeneity but still share similarities among patients.</p>
</div>
<div id="simulation-setup" class="section level2">
<h2>Simulation setup</h2>
<pre class="r"><code>library(Matrix)
library(ebnm)
library(flashier)
library(pheatmap)
library(gridExtra)
library(tidyverse)

### load in simulated loadings and factors
L &lt;- readRDS(&quot;data/model.rds&quot;)$L
F &lt;- readRDS(&quot;data/model.rds&quot;)$F
F[,1] &lt;- F[,1] - 10

### plot the poplulation factorization of truth, where F is orthogonal across columns 
L.pop &lt;- L[, 1:9]
F.pop &lt;- F[, 1:9]
F.pop[,2:5] &lt;- F[,2:5] + F[,10]
F.pop[,6:9] &lt;- F[,6:9] + F[,11]
plt.L &lt;- pheatmap(L.pop[,-1], show_rownames = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, 
                  main = &quot;True L\n population factorization&quot;, silent = TRUE)
plt.F &lt;- pheatmap(F.pop[,-1], show_rownames = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, 
                  main = &quot;True F\n population factorization&quot;, silent = TRUE)
grid.arrange(plt.L$gtable, plt.F$gtable, ncol=2)</code></pre>
<p><img src="figure/presentation.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### plot the drift factorization of truth, where F is orthogonal across columns 
plt.L &lt;- pheatmap(L[,-1], show_rownames = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, main = &quot;True L\n drift factorization&quot;, silent = TRUE)
plt.F &lt;- pheatmap(F[,-1], show_rownames = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, main = &quot;True F\n drift factorization&quot;, silent = TRUE)
grid.arrange(plt.L$gtable, plt.F$gtable, ncol=2)</code></pre>
<p><img src="figure/presentation.Rmd/unnamed-chunk-1-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>### simulate the data
set.seed(100)
X &lt;- L %*% t(F) + matrix(rnorm(800*2000, 0, 1), nrow=800, ncol=2000)
XXt &lt;- X %*% t(X)/ncol(X)</code></pre>
<div id="define-functions-to-fit-snn" class="section level3">
<h3>Define functions to fit SNN</h3>
<pre class="r"><code>source(&quot;code/functions.R&quot;)</code></pre>
</div>
<div id="define-plot-functions" class="section level3">
<h3>Define plot functions</h3>
<pre class="r"><code># plot posterior mean of L and F after rescaling
plotLF &lt;- function(fit, title){
  fit.L &lt;- fit$loadings.pm[[1]][, fit$pve &gt; 0]
  fit.L &lt;- t(t(fit.L)/apply(fit.L, 2, max))
  colnames(fit.L) &lt;- paste0(&quot;k&quot;, 1:ncol(fit.L))
  fit.F &lt;- fit$loadings.pm[[2]][, fit$pve &gt; 0]
  fit.F &lt;- t(t(fit.F)/apply(fit.F, 2, max))
  colnames(fit.F) &lt;- paste0(&quot;k&quot;, 1:ncol(fit.F))
  plt.L &lt;- pheatmap(fit.L, cluster_rows = FALSE, cluster_cols = FALSE, silent = TRUE, main=title[1])
  plt.F &lt;- pheatmap(fit.F, cluster_rows = FALSE, cluster_cols = FALSE, silent = TRUE, main=title[2])
  grid.arrange(plt.L$gtable, plt.F$gtable, ncol=2)
}</code></pre>
</div>
<div id="fit-snn-to-the-data-matrix" class="section level3">
<h3>Fit SNN to the data matrix</h3>
<p>As can be seen, SNN fit to the cell <span class="math inline">\(\times\)</span> gene data matrix (point-exponential prior on loadings, point-Laplace prior on factors) yields a population factorization, in which the loadings fail to reveal the similarities among patients.</p>
<pre class="r"><code>### initialize by running flash without constraints on L
fit.pn &lt;- flash.init(X, S = 1/sqrt(nrow(X)), var.type = 2) %&gt;% flash.add.greedy(Kmax = 11, prior.family = prior.point.laplace()) %&gt;% flash.backfit()</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Factor doesn&#39;t significantly increase objective and won&#39;t be added.
Wrapping up...
Done.
Backfitting 8 factors (tolerance: 2.38e-02)...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.</code></pre>
<pre class="r"><code>snn.init &lt;- init.snn.LL(fit.pn)

### run SNN with point exponential prior on L
fit.snn &lt;- flash.init(X, S = 1/sqrt(nrow(X)), var.type = 2
) %&gt;% flash.init.factors(EF = snn.init, prior.family = c(as.prior(ebnm::ebnm_point_exponential, sign = 1), prior.point.laplace())
) %&gt;% flash.backfit()</code></pre>
<pre><code>Backfitting 15 factors (tolerance: 2.38e-02)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.</code></pre>
<pre class="r"><code>### plot the estimated L
fit.L &lt;- fit.snn$loadings.pm[[1]][, fit.snn$pve &gt; 0]
fit.L &lt;- t(t(fit.L)/apply(fit.L, 2, max))
colnames(fit.L) &lt;- paste0(&quot;k&quot;, 1:ncol(fit.L))
pheatmap(fit.L, cluster_rows = FALSE, cluster_cols = FALSE, main=&quot;L from SNN fit to data matrix&quot;)</code></pre>
<p><img src="figure/presentation.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="fit-snn-to-the-covariance-matrix-with-a-point-exponential-prior" class="section level3">
<h3>Fit SNN to the covariance matrix with a point-exponential prior</h3>
<p>To encourage a drift factorization in which the loadings reveal similarities among patients and are thus more informative, we fit SNN to the cell <span class="math inline">\(\times\)</span> cell covariance matrix to estimate the loadings first, and then fit SNN to the cell <span class="math inline">\(\times\)</span> gene data matrix by keeping loadings fixed.</p>
<pre class="r"><code>### fit unconstrained flash with point Laplace prior to XXt while ignoring the diagonal component for now
fit.cov &lt;- flash.init(XXt, var.type = 0) %&gt;% flash.add.greedy(Kmax = 11, prior.family = prior.point.laplace()) %&gt;% flash.backfit()</code></pre>
<pre><code>Adding factor 1 to flash object...
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Adding factor 10 to flash object...
Adding factor 11 to flash object...
Wrapping up...
Done.
Backfitting 11 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.</code></pre>
<pre class="r"><code>### fit unconstrained flash again with the diagonal component
fit.cov &lt;- fit.ebcovmf(dat=XXt, fl=fit.cov, prior=prior.point.laplace())$fl</code></pre>
<pre><code>Backfitting 11 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.
Backfitting 11 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.</code></pre>
<pre class="r"><code>### initialize the nonnegative fit based on the unconstrained flash fit 
snn.cov.init &lt;- init.snn.cov(fit.cov)

### fit flash with point exponential prior to XXt while ignoring the diagonal component for now
fit.cov &lt;- flash.init(XXt, var.type = 0) %&gt;% flash.init.factors(EF=snn.cov.init, prior.family=as.prior(ebnm::ebnm_point_exponential, sign = 1)) 

### fit flash again with the diagonal component
fit.cov &lt;- fit.ebcovmf(dat=XXt, fl=fit.cov, prior=as.prior(ebnm::ebnm_point_exponential, sign = 1))$fl</code></pre>
<pre><code>Backfitting 21 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.
Backfitting 16 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.</code></pre>
<pre class="r"><code>### plot the estimated L and F
plotLF(fit.cov, title=c(&quot;L from SNN fit to covariance\n point exponential prior)&quot;, &quot;F from SNN fit to covariance\n point exponential prior)&quot;))</code></pre>
<p><img src="figure/presentation.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="fit-snn-to-the-covariance-matrix-with-a-generalized-binary-prior" class="section level3">
<h3>Fit SNN to the covariance matrix with a generalized binary prior</h3>
<p>In the SNN fit to the covariance matrix with a point-exponential prior, the loadings do reveal similaries among patients, but the solution still looks quite noisy and is not close to the drift factorization desired.</p>
<p>To fix this, we instead fit SNN to the covariance matrix with a generalized binary prior, which can be expressed as <span class="math display">\[\begin{align}
g \sim (1-\pi) \delta_0 + \pi N_{[0, \infty)}(\mu, \sigma^2),
\end{align}\]</span> where the truncated normal component is highly spiky, i.e., <span class="math inline">\(\frac{\mu}{\sigma}\)</span> is large.</p>
<pre class="r"><code>### fit flash with generalized binary prior to XXt while ignoring the diagonal component for now
fit.cov &lt;- flash.init(XXt, var.type = 0) %&gt;% flash.init.factors(EF=snn.cov.init, prior.family=as.prior(ebnm_binary_general, sign = 1)) 

### fit flash again with the diagonal component
fit.cov &lt;- fit.ebcovmf(dat=XXt, fl=fit.cov, prior=as.prior(ebnm_binary_general, sign = 1))$fl</code></pre>
<pre><code>Backfitting 21 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e+03...
  Difference between iterations is within 1.0e+02...
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.
Backfitting 12 factors (tolerance: 9.54e-03)...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.</code></pre>
<pre class="r"><code>### plot the estimated L and F
plotLF(fit.cov, title=c(&quot;L from SNN fit to covariance\n general binary prior)&quot;, &quot;F from SNN fit to covariance\n general binary prior)&quot;))</code></pre>
<p><img src="figure/presentation.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.5.1   stringr_1.4.0   dplyr_1.0.6     purrr_0.3.4    
 [5] readr_2.0.2     tidyr_1.1.3     tibble_3.1.2    ggplot2_3.3.4  
 [9] tidyverse_1.3.1 gridExtra_2.3   pheatmap_1.0.12 flashier_0.2.7 
[13] ebnm_0.1-52     Matrix_1.3-4    workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] httr_1.4.2         sass_0.4.0         jsonlite_1.7.2     splines_4.1.0     
 [5] modelr_0.1.8       bslib_0.2.5.1      assertthat_0.2.1   horseshoe_0.2.0   
 [9] highr_0.9          mixsqp_0.3-43      deconvolveR_1.2-1  cellranger_1.1.0  
[13] yaml_2.2.1         pillar_1.6.1       backports_1.2.1    lattice_0.20-44   
[17] glue_1.4.2         digest_0.6.27      RColorBrewer_1.1-2 promises_1.2.0.1  
[21] rvest_1.0.2        colorspace_2.0-1   htmltools_0.5.1.1  httpuv_1.6.1      
[25] pkgconfig_2.0.3    invgamma_1.1       broom_0.7.9        haven_2.4.3       
[29] scales_1.1.1       whisker_0.4        later_1.2.0        tzdb_0.1.2        
[33] git2r_0.28.0       generics_0.1.0     ellipsis_0.3.2     withr_2.4.2       
[37] ashr_2.2-47        cli_3.0.1          readxl_1.3.1       magrittr_2.0.1    
[41] crayon_1.4.1       evaluate_0.14      fs_1.5.0           fansi_0.5.0       
[45] xml2_1.3.2         truncnorm_1.0-8    tools_4.1.0        hms_1.1.0         
[49] REBayes_2.2        lifecycle_1.0.0    reprex_2.0.1       trust_0.1-8       
[53] munsell_0.5.0      irlba_2.3.3        compiler_4.1.0     jquerylib_0.1.4   
[57] rlang_0.4.11       grid_4.1.0         rstudioapi_0.13    rmarkdown_2.11    
[61] gtable_0.3.0       DBI_1.1.1          R6_2.5.0           lubridate_1.8.0   
[65] knitr_1.33         utf8_1.2.1         rprojroot_2.0.2    stringi_1.6.2     
[69] parallel_4.1.0     SQUAREM_2021.1     Rcpp_1.0.7         vctrs_0.3.8       
[73] dbplyr_2.1.1       tidyselect_1.1.1   xfun_0.25         </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
